### PR TITLE
Pipeline wizard

### PR DESCRIPTION
**Pull request by sudhirverma**
Fix: #310

### CODE REVIEW

**Comment by sudhirverma** [comment]
Workspace UI @siamaksade and @evidolob 
Screenshot 2020-07-15 at 1 34 44 AM

**Comment by sudhirverma** [comment]
Parameters, Git and Image Resource UI 

Screenshot 2020-07-15 at 1 35 44 AM

**Comment by evidolob** [comment]
@sudhirverma I got:
Screenshot 2020-07-24 at 09 53 19

**Comment by evidolob** [comment]
@sudhirverma I have empty wizard for [this pipeline](https://raw.githubusercontent.com/tektoncd/pipeline/master/examples/v1beta1/pipelineruns/clustertask-pipelinerun.yaml)
Screenshot 2020-07-27 at 09 52 28

File: src/pipeline/wizard.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    out.push(`<script
+				src="${escapeAttribute(this.editor.webview.asWebviewUri(vscode.Uri.file(path.join(contextGlobalState.extensionPath, '/out/webview/pipeline/index.js'))))}"
+				nonce="${nonce}"
+				charset="UTF-8"></script>`);
+    return out.join('\n');
+  }
+
+  private getImagesUri(): { [key: string]: string } {
+    const result: { [key: string]: string } = Object.create(null);
+    result['task'] = this.editor.webview.asWebviewUri(vscode.Uri.file(path.join(contextGlobalState.extensionPath, 'images', 'T.svg'))).toString();
```
Comment (evidolob) [inline src/pipeline/wizard.ts:154]: Do we use that?

File: src/webview/pipeline/app/common/item.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+import { SelectWidget } from '../element/selectwidget';
+import { EditItem } from '../element/maincontent';
+import { createDiv } from '../utils/util';
+import { InputWidget } from '../element/inputwidget';
+import { ButtonsPanel } from '../element/buttonspanel';
+import { PipelineStart, Trigger } from './types';
+import { disableButton } from '..';
+
+export function disableRemoveButton(event: Node & ParentNode): void {
```
Comment (evidolob) [inline src/webview/pipeline/app/common/item.ts:14]: I think that should be member of the `Button` class

File: src/webview/pipeline/app/common/item.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  const selectedItem = event.querySelectorAll('[id^=items-section-workspace-new-item]');
+  if (selectedItem.length === 1) {
+    selectedItem[0].lastElementChild.firstElementChild.className = 'close-button-disable';
+  } else {
+    selectedItem[0].lastElementChild.firstElementChild.className = 'close-button';
+  }
+}
+
+// Disable button if section are not selected for workspace
+export function disableSelection(nodeList: HTMLCollectionOf<HTMLSelectElement>): boolean {
```
Comment (evidolob) [inline src/webview/pipeline/app/common/item.ts:24]: Same there, for selector class

File: src/webview/pipeline/app/common/item.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    if (findOption) {
+      startButton.className = 'startButton-disable';    // Disable the button.
+      return false;
+    } else {
+      startButton.className = 'startButton';
+    }
+  }
+}
+
+export function createItem(event: Node & ParentNode, optionId: string, selectValue?: string, initialValue?: PipelineStart, trigger?: Trigger): void {
```
Comment (evidolob) [inline src/webview/pipeline/app/common/item.ts:41]: I think, name of this function do not represent what id do. Create element and add it to parent it two different operation.

File: src/webview/pipeline/app/common/resource.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
+/*-----------------------------------------------------------------------------------------------
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE file in the project root for license information.
+ *-----------------------------------------------------------------------------------------------*/
+
+import { PipelineStart } from './types';
+
+
+export function parameter(paramName: string, defaultValue: string, initialValue: PipelineStart): void {
```
Comment (evidolob) [inline src/webview/pipeline/app/common/resource.ts:9]: Name of this function is not clear.

File: src/webview/pipeline/app/common/resource.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        return true;
+      }
+    });
+    if (!found) {
+      initialValue.params.push({name: paramName, default: defaultValue});
+    }
+  }
+}
+
+export function createResourceJson(resourceName: string, resourceReference: string, initialValue: PipelineStart): void {
```
Comment (evidolob) [inline src/webview/pipeline/app/common/resource.ts:25]: Same name problem, function named `create*` but return `void`

File: src/webview/pipeline/app/common/resource.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        return true;
+      }
+    });
+    if (!found) {
+      initialValue.resources.push({name: resourceName, resourceRef: resourceReference});
+    }
+  }
+}
+
+export function createWorkspaceJson(resourceName: string, workspaceResourceType: string, initialValue: PipelineStart, workspaceResourceName?: string, keyName?: string, valueName?: string): void {
```
Comment (evidolob) [inline src/webview/pipeline/app/common/resource.ts:41]: Same there.

File: src/webview/pipeline/app/element/selectwidget.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        if (selectedItem.length) {
+          selectedItem.forEach(element => element.remove());
+          buttonItem.forEach(element => element.remove());
+          createItem(event, optionId, select.value, this.initialValue, this.trigger);
+        } else {
+          createItem(event, optionId, select.value, this.initialValue, this.trigger);
+        }
+      }
+    } catch (err) {
+      // ignores
```
Comment (evidolob) [inline src/webview/pipeline/app/element/selectwidget.ts:83]: What error can be catch there?

Comment (sudhirverma) [reply src/webview/pipeline/app/element/selectwidget.ts:83]: In this try-catch I check the workspace content ex: config map, secret, and PVC, if it is not workspace content then it throws error so I am ignoring that error here.

File: src/pipeline/wizard.ts (hunk 1 above)
Comment (sudhirverma) [reply src/pipeline/wizard.ts:154]: I am not using this code I will remove this code.

File: src/webview/pipeline/app/common/item.ts (hunk 2 above)
Comment (sudhirverma) [reply src/webview/pipeline/app/common/item.ts:14]: I am using this function in two different files, I will move this function in a different file. I will create disablebutton file and move the function disableRemoveButton and disableSelection in that file.

File: src/webview/pipeline/app/common/resource.ts (hunk 6 above)
Comment (sudhirverma) [reply src/webview/pipeline/app/common/resource.ts:25]: I am storing the resource data here I will change the function name accordingly.

File: src/webview/pipeline/app/common/item.ts (hunk 4 above)
Comment (sudhirverma) [reply src/webview/pipeline/app/common/item.ts:41]: I will create two different functions one for creating the item and another for adding items.

**Comment by sudhirverma** [comment]
> @sudhirverma I have empty wizard for [this pipeline](https://raw.githubusercontent.com/tektoncd/pipeline/master/examples/v1beta1/pipelineruns/clustertask-pipelinerun.yaml)
> Screenshot 2020-07-27 at 09 52 28

if YAML doesn't have params, git or image resources, and workspace I will directly start the pipeline. without showing the wizard.

**Comment by sudhirverma** [comment]
@siamaksade and @evidolob 

param, git and Image Resource Gif

resource


Workspace Gif

workspace

**Comment by evidolob** [comment]
@sudhirverma When wizard starts pipeline, it is open pipeline run diagram for previous pipelinerun:
Screenshot 2020-07-30 at 11 10 50

File: src/tekton/pipeline.ts [hunk 9]
```diff
@@ -6,45 +6,47 @@
...
-      serviceAcct: value.spec.serviceAccount ? value.spec.serviceAccount : undefined
-    })).filter((obj) => obj.name === pipeline.getName());
-    const inputStartPipeline = await PipelineContent.startObject(pipelineTrigger, 'Pipeline');
+      const pipelineTrigger = data.map<Trigger>(value => ({
+        name: value.metadata.name,
+        resources: value.spec.resources,
+        params: value.spec.params ? value.spec.params : undefined,
+        workspaces: value.spec['workspaces'] ? value.spec['workspaces'] : undefined,
+        serviceAcct: value.spec.serviceAccount ? value.spec.serviceAccount : undefined
+      })).filter((obj) => obj.name === pipeline.getName());
```
Comment (evidolob) [inline src/tekton/pipeline.ts:43]: It is more efficient to filter first that map

**Comment by sudhirverma** [comment]
> @sudhirverma When wizard starts pipeline, it is open pipeline run diagram for previous pipelinerun:
> Screenshot 2020-07-30 at 11 10 50

okay, I will check this out.

File: src/webview/pipeline/app/index.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE file in the project root for license information.
+ *-----------------------------------------------------------------------------------------------*/
+
+import './style.css';
+import 'vscode-codicons/dist/codicon.ttf'
+import { PipelineRunEditor } from './editor';
+import { Trigger } from './utils/types';
+
+declare let acquireVsCodeApi: any;
```
Comment (evidolob) [inline src/webview/pipeline/app/index.ts:11]: Please fix `Unexpected any. Specify a different type`

File: src/webview/pipeline/app/widgets/selectwidget.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        input.removeAttribute('disabled');
+        input.id = 'enabled';
+        input.title = '';
+        input.parentNode.parentElement.className = 'editor-input-box';
+        event.parentNode.parentNode.querySelector('select').firstElementChild.remove();
+      }
+    }
+  }
+
+  createWorkspaceElement(event: Node & ParentNode, select: HTMLSelectElement): void {
```
Comment (evidolob) [inline src/webview/pipeline/app/widgets/selectwidget.ts:66]: This is quite big method, can you split it in to several?

File: src/tkn.ts [hunk 12]
```diff
@@ -876,6 +902,7 @@ export interface Tkn {
   getPipelineRunsList(pipelineRun?: TektonNode): Promise<TektonNode[]>;
   getTaskRunList(taskRun?: TektonNode): Promise<TektonNode[]>;
   getRawPipelineRun(name: string): Promise<PipelineRunData | undefined>;
+  _getPipelineRuns(pipeline: TektonNode, pipelineName?: string): Promise<TektonNode[]> | undefined;
```
Comment (evidolob) [inline src/tkn.ts:905]: Why do you have `_` in function name? Maybe better to find more descriptive name for that function.

File: src/webview/pipeline/app/widgets/inputwidget.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE file in the project root for license information.
+ *-----------------------------------------------------------------------------------------------*/
+
+import { createDiv } from '../utils/util';
+import { BaseWidget } from './widget';
+import { PipelineStart } from '../utils/types';
+import { TknResourceType } from '../utils/const';
+import { collectParameterData, collectResourceData } from '../utils/resource';
+import { disableButton } from '..';
```
Comment (evidolob) [inline src/webview/pipeline/app/widgets/inputwidget.ts:11]: This import looks suspicious to me, please import proper file, not entire directory

**Comment by sudhirverma** [comment]
@evidolob done the changes

**Pull request merged**